File: typescript/questions/subarraySum.ts
function subarraySum(nums: number[], k: number): number {
  let nl = nums.length;
  let mp: Map<number, number> = new Map();
  let ps = 0;
  let ans = 0;
  mp.set(ps, (mp.get(ps) ?? 0) + 1);
  for (let p = 0; p < nl; ++p) {
    ps += nums[p];
    ans += mp.get(ps - k) ?? 0;
    mp.set(ps, (mp.get(ps) ?? 0) + 1);
  }
  return ans;
}

File: rust/questions/subarray_sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut mp: HashMap<i32, i32> = HashMap::new();
        let mut ans = 0;
        let mut ps = 0;
        *mp.entry(ps).or_insert(0) += 1;
        for p in 0..nums.len() {
            ps += nums[p];
            if let Some(&c) = mp.get(&(ps - k)) {
                ans += c;
            }
            *mp.entry(ps).or_insert(0) += 1;
        }
        ans
    }
}
